### ISSUE TITLE
security: maker can lose funds due to a crash

### ISSUE THREAD

**Issue by moshababo** [ExchangeUnion/xud#1050]
Scenario:
* taker is paying the maker the 1st leg of the swap.
* maker receives the taker HTLC, and is paying the taker the 2nd leg of the swap.
* taker receives the maker HTLC, and settles his invoice as expected.
* maker 2nd leg payment completed and he receives the preimage, but his xud crashes before he manages to settle his own invoice.

Outcome: maker already paid but cannot redeem his pending payment because he lost the preimage. Taker funds are still locked but he can get them back after the HTLC will expire, leaving the maker with none.

**Comment by sangaman** [comment]
As we discussed, the solution here is to write the preimage and hash for every pending swap to database as soon as they become known to us. If xud crashes, when it restarts we check the database for pending swaps and try to settle any invoices for which we known the preimage. If we don't have the preimage, we call `ListPayments` on lnd to see if we've made a payment for any of the pending hashes, and thus learned of the preimage for that hash, and then try to settle an invoice using that preimage. If we don't have the preimage after all that, we cancel any pending invoices to avoid funds from being lucked up unnecessarily.
